### PR TITLE
Secrets manager resource init

### PR DESCRIPTION
**Pull request by thisisshi**
Related to: https://github.com/capitalone/cloud-custodian/issues/2390

### CODE REVIEW

File: c7n/resources/secretsmanager.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        client = local_session(self.session_factory).client('secretsmanager')
+
+        def _augment(r):
+            r['Versions'] = client.list_secret_version_ids(SecretId=r['ARN']).get('Versions')
+            self.log.info('adding version info')
+            return r
+
+        resources = super(SecretsManager, self).augment(resources)
+
+        with self.executor_factory(max_workers=1) as w:
```
Comment (kapilt) [inline c7n/resources/secretsmanager.py]: if your just doing concurrenty one, can just turn this into a loop.

File: c7n/resources/secretsmanager.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        client = local_session(self.manager.session_factory).client('secretsmanager')
+        for r in resources:
+            r['Tags'].extend([{'Key': t['Key'], 'Value':t['Value']} for t in tags])
+            try:
+                client.tag_resource(SecretId=r['ARN'], Tags=r['Tags'])
+            except ClientError as e:
+                if e['Error']['Code'] == 'InternalFailure':
+                    self.log.error(
+                        'Exception when tagging resource %s' %
+                        (r['ARN'], e['Error']['Code']))
```
Comment (kapilt) [inline c7n/resources/secretsmanager.py]: this is silently dropping all other errors.

File: c7n/resources/secretsmanager.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    permissions = ('secretsmanager:TagResource',)
+
+    def process_resource_set(self, resources, tags):
+        client = local_session(self.manager.session_factory).client('secretsmanager')
+        for r in resources:
+            r['Tags'].extend([{'Key': t['Key'], 'Value':t['Value']} for t in tags])
+            try:
+                client.tag_resource(SecretId=r['ARN'], Tags=r['Tags'])
+            except ClientError as e:
+                if e['Error']['Code'] == 'InternalFailure':
```
Comment (kapilt) [inline c7n/resources/secretsmanager.py]: internal failures should be re-raising.

File: c7n/resources/secretsmanager.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                    self.log.error('Exception when tagging resource %s' %
+                        (['ARN'], e['Error']['Code']))
```
Comment (kapilt) [inline c7n/resources/secretsmanager.py]: other errors are silently passing here.

File: c7n/resources/secretsmanager.py (hunk 1 above)
Comment (thisisshi) [reply c7n/resources/secretsmanager.py]: 🤦‍♂️ oops, I was debating turning up the concurrency but decided against it and never switched it back

File: c7n/resources/secretsmanager.py (hunk 3 above)
Comment (thisisshi) [reply c7n/resources/secretsmanager.py]: This is a result of a bug with tag name conflicts it _should_ be a client error, snippet to reproduce:

```python
client = boto3.client('secretsmanager')
client.tag_resource(SecretId="test", Tags=[{"Key":"test", "Value":"value"},{"Key":"test", "Value":"value"}])

a ticket's been open with AWS for now

Comment (kapilt) [reply c7n/resources/secretsmanager.py]: yeah, for now i suggest we not catch here then, the internal failure should hopefully get fixed by the service team. ie. those errors need human attention. switching how you pass the tags should also resolve.

File: c7n/resources/secretsmanager.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                  key: tag-key
+                  value: tag-value
+    """
+
+    permissions = ('secretsmanager:TagResource',)
+
+    def process_resource_set(self, resources, tags):
+        client = local_session(self.manager.session_factory).client('secretsmanager')
+        for r in resources:
+            r['Tags'].extend([{'Key': t['Key'], 'Value':t['Value']} for t in tags])
```
Comment (kapilt) [inline c7n/resources/secretsmanager.py]: you should convert to a dict here, so you don't have duplicate keys, and then generate the list when invoking.

File: tests/test_secretsmanager.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import absolute_import, division, print_function, unicode_literals
+from .common import BaseTest, functional
+
+import time
+
+class TestSecretsManager(BaseTest):
+
+    def test_secrets_manager_resource(self):
```
Comment (kapilt) [inline tests/test_secretsmanager.py]: there's a lot of recorded json on this pr, this particular test feels a bit redundant and could be dropped to slim it down.

File: tests/test_secretsmanager.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            'name': 'secrets-manager-resource',
+            'resource': 'secrets-manager'}, session_factory=session)
+        resources = p.run()
+        self.assertEqual(len(resources), 1)
+        self.assertTrue(resources[0].get('Versions'))
+
+    def test_secrets_manager_tag_resource(self):
+        session = self.replay_flight_data('test_tag_secrets_manager_resource')
+        client = session(region='us-east-1').client('secretsmanager')
+        og_tags = client.describe_secret(
```
Comment (kapilt) [inline tests/test_secretsmanager.py]: this one could just be converted to a value filter in the policy. `Tags: empty`

File: c7n/resources/secretsmanager.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+from c7n.query import QueryResourceManager
+from c7n.tags import RemoveTag, Tag, TagActionFilter, TagDelayedAction
+from c7n.utils import local_session
+
+
+@resources.register('secrets-manager')
+class SecretsManager(QueryResourceManager):
+    filters = FilterRegistry('secrets-manager.filters')
+    filters.register('marked-for-op', TagActionFilter)
+    filter_registry = filters
```
Comment (kapilt) [inline c7n/resources/secretsmanager.py]: this assignment could be omitted by just making the `filter_registry = FilterRegistry(` above

File: c7n/resources/secretsmanager.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        id = 'ARN'
+        name = 'Name'
+        dimension = None
+        filter_name = None
+
+    def augment(self, resources):
+        client = local_session(self.session_factory).client('secretsmanager')
+
+        def _augment(r):
+            r['Versions'] = client.list_secret_version_ids(SecretId=r['ARN']).get('Versions')
```
Comment (kapilt) [inline c7n/resources/secretsmanager.py]: do we need the versions by default?

Comment (thisisshi) [reply c7n/resources/secretsmanager.py]: So the only way to get created/modified dates are by grabbing versions explicitly. At this point I'm leaning more towards a dedicated filter to handle this as secrets can have a large number of versions, especially if users are practicing secret rotation frequently. Plus, there needs to be some additional work to be done regarding handling the paginated results from AWS.

**Pull request merged**